File: js-service/__tests__/circuit-breaker.test.ts
import { describe, it, expect, jest, beforeEach, afterEach } from '@jest/globals'
import {
  CircuitBreaker,
  CircuitBreakerOpenError,
  CircuitState,
  DistributedCircuitBreakerClient,
  withCircuitBreaker,
} from '../src/circuit-breaker'

describe('CircuitBreaker', () => {
  afterEach(() => {
    jest.clearAllMocks()
    jest.useRealTimers()
  })

  it('execute resolves and updates metrics on success', async () => {
    const breaker = new CircuitBreaker('svc-success')
    const result = await breaker.execute(async () => 'ok')

    expect(result).toBe('ok')
    const info = breaker.getHealthInfo()
    expect(info.state).toBe(CircuitState.CLOSED)
    expect(info.metrics.totalCalls).toBe(1)
    expect(info.metrics.successfulCalls).toBe(1)
    expect(info.metrics.failedCalls).toBe(0)
    expect(info.metrics.lastSuccessTime).not.toBeNull()
    expect(typeof info.metrics.averageResponseTimeMs).toBe('number')
  })

  it('execute rejects and updates metrics on failure, remains CLOSED below threshold', async () => {
    const breaker = new CircuitBreaker('svc-fail', { failureThreshold: 3 })
    await expect(
      breaker.execute(async () => {
        throw new Error('boom')
      })
    ).rejects.toThrow('boom')

    const info = breaker.getHealthInfo()
    expect(info.state).toBe(CircuitState.CLOSED)
    expect(info.metrics.totalCalls).toBe(1)
    expect(info.metrics.failedCalls).toBe(1)
    expect(info.metrics.lastFailureTime).not.toBeNull()
  })

  it('transitions to OPEN after reaching failureThreshold', async () => {
    const breaker = new CircuitBreaker('svc-open', { failureThreshold: 2 })
    await expect(
      breaker.execute(async () => {
        throw new Error('fail1')
      })
    ).rejects.toThrow('fail1')
    await expect(
      breaker.execute(async () => {
        throw new Error('fail2')
      })
    ).rejects.toThrow('fail2')

    expect(breaker.getState()).toBe(CircuitState.OPEN)
    expect(breaker.getHealthInfo().metrics.stateTransitions).toBe(1)
  })

  it('rejects when OPEN without fallback and increments rejectedCalls', async () => {
    const breaker = new CircuitBreaker('svc-open-no-fallback', { failureThreshold: 1, timeoutMs: 1000 })
    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow('fail')

    await expect(
      breaker.execute(async () => 'should-not-run')
    ).rejects.toBeInstanceOf(CircuitBreakerOpenError)

    const info = breaker.getHealthInfo()
    expect(info.metrics.rejectedCalls).toBe(1)
    expect(info.metrics.totalCalls).toBe(1)
  })

  it('OPEN with fallback returns fallback and increments rejectedCalls', async () => {
    const breaker = new CircuitBreaker('svc-open-fallback', { failureThreshold: 1, timeoutMs: 1000 })
    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow('fail')

    const result = await breaker.execute(async () => 'nope', async () => 'fallback')
    expect(result).toBe('fallback')

    const info = breaker.getHealthInfo()
    expect(info.metrics.rejectedCalls).toBe(1)
    expect(info.metrics.totalCalls).toBe(1)
  })

  it('OPEN error message includes name and remaining time', async () => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date(0))
    const breaker = new CircuitBreaker('svc-open-err-msg', { failureThreshold: 1, timeoutMs: 1000 })

    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow()

    jest.setSystemTime(new Date(100))
    await expect(
      breaker.execute(async () => 'x')
    ).rejects.toThrowError(new RegExp(`Circuit breaker 'svc-open-err-msg' is open\\. Retry after 900ms`))
  })

  it('transitions to HALF_OPEN after timeout from OPEN on getState()', async () => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date(0))
    const breaker = new CircuitBreaker('svc-half-open', { failureThreshold: 1, timeoutMs: 500 })

    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow()

    expect(breaker.getState()).toBe(CircuitState.OPEN)

    jest.setSystemTime(new Date(600))
    expect(breaker.getState()).toBe(CircuitState.HALF_OPEN)
  })

  it('HALF_OPEN allows limited calls and rejects further calls', async () => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date(0))
    const breaker = new CircuitBreaker('svc-half-open-limit', {
      failureThreshold: 1,
      timeoutMs: 100,
      halfOpenMaxCalls: 2,
      successThreshold: 10,
    })

    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow()

    jest.setSystemTime(new Date(200))
    expect(breaker.getState()).toBe(CircuitState.HALF_OPEN)

    const r1 = await breaker.execute(async () => 'ok1')
    const r2 = await breaker.execute(async () => 'ok2')
    expect(r1).toBe('ok1')
    expect(r2).toBe('ok2')

    const r3 = await breaker.execute(async () => 'no', async () => 'fb')
    expect(r3).toBe('fb')

    const info = breaker.getHealthInfo()
    expect(info.state).toBe(CircuitState.HALF_OPEN)
    expect(info.metrics.rejectedCalls).toBe(1)
  })

  it('HALF_OPEN closes after reaching successThreshold', async () => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date(0))
    const breaker = new CircuitBreaker('svc-half-open-close', {
      failureThreshold: 1,
      timeoutMs: 100,
      halfOpenMaxCalls: 5,
      successThreshold: 2,
    })

    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow()

    jest.setSystemTime(new Date(200))
    expect(breaker.getState()).toBe(CircuitState.HALF_OPEN)

    await breaker.execute(async () => 'ok1')
    await breaker.execute(async () => 'ok2')

    expect(breaker.getState()).toBe(CircuitState.CLOSED)
    const info = breaker.getHealthInfo()
    expect(info.failureCount).toBe(0)
    expect(info.metrics.stateTransitions).toBe(2) // OPEN -> HALF_OPEN -> CLOSED
  })

  it('HALF_OPEN failure transitions back to OPEN', async () => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date(0))
    const breaker = new CircuitBreaker('svc-half-open-failure', {
      failureThreshold: 1,
      timeoutMs: 100,
      halfOpenMaxCalls: 3,
      successThreshold: 5,
    })

    await expect(
      breaker.execute(async () => {
        throw new Error('fail')
      })
    ).rejects.toThrow()

    jest.setSystemTime(new Date(200))
    expect(breaker.getState()).toBe(CircuitState.HALF_OPEN)

    await expect(
      breaker.execute(async () => {
        throw new Error('half-open-fail')
      })
    ).rejects.toThrow('half-open-fail')

    expect(breaker.getState()).toBe(CircuitState.OPEN)
  })

  it('sliding window failure rate triggers OPEN before failureThreshold', async () => {
    const breaker = new CircuitBreaker('svc-sliding', {
      failureThreshold: 100,
      slidingWindowSize: 4,
      failureRateThreshold: 0.5,
    })

    // 3 consecutive failures -> window has [false,false,false,true] => 75% failures
    await expect(breaker.execute(async () => { throw new Error('f1') })).rejects.toThrow()
    await expect(breaker.execute(async () => { throw new Error('f2') })).rejects.toThrow()
    await expect(breaker.execute(async () => { throw new Error('f3') })).rejects.toThrow()

    expect(breaker.getState()).toBe(CircuitState.OPEN)
  })

  it('executeSync handles success and updates metrics', () => {
    const breaker = new CircuitBreaker('svc-sync-success')
    const result = breaker.executeSync(() => 42)
    expect(result).toBe(42)
    const info = breaker.getHealthInfo()
    expect(info.metrics.totalCalls).toBe(1)
    expect(info.metrics.successfulCalls).toBe(1)
    expect(info.state).toBe(CircuitState.CLOSED)
  })

  it('executeSync throws on failure and updates metrics', () => {
    const breaker = new CircuitBreaker('svc-sync-failure', { failureThreshold: 5 })
    expect(() => breaker.executeSync(() => { throw new Error('sync-fail') })).toThrow('sync-fail')
    const info = breaker.getHealthInfo()
    expect(info.metrics.failedCalls).toBe(1)
    expect(info.state).toBe(CircuitState.CLOSED)
  })

  it('getHealthInfo returns correct structure and failureRate', async () => {
    const breaker = new CircuitBreaker('svc-health', {
      slidingWindowSize: 4,
      failureThreshold: 10,
      failureRateThreshold: 0.9,
    })

    await breaker.execute(async () => 's1')
    await expect(breaker.execute(async () => { throw new Error('f1') })).rejects.toThrow()
    await breaker.execute(async () => 's2')
    await expect(breaker.execute(async () => { throw new Error('f2') })).rejects.toThrow()

    const info = breaker.getHealthInfo()
    expect(info.name).toBe('svc-health')
    expect(info.state).toBe(CircuitState.CLOSED)
    expect(info.failureRate).toBeCloseTo(0.5, 5)
    expect(info.config.failureThreshold).toBeDefined()
    expect(info.config.successThreshold).toBeDefined()
    expect(info.config.timeoutMs).toBeDefined()
    // Ensure only partial config exposed
    expect(Object.keys(info.config).sort()).toEqual(['failureThreshold', 'successThreshold', 'timeoutMs'].sort())
  })

  it('getOrCreate returns same instance for same name and registry contains it', () => {
    const b1 = CircuitBreaker.getOrCreate('shared', { failureThreshold: 2 })
    const b2 = CircuitBreaker.getOrCreate('shared', { failureThreshold: 10 })
    expect(b2).toBe(b1)

    const registry = CircuitBreaker.getRegistry()
    expect(registry.has('shared')).toBe(true)

    // Returned registry is a copy; modifications don't affect internal
    registry.delete('shared')
    const registry2 = CircuitBreaker.getRegistry()
    expect(registry2.has('shared')).toBe(true)
  })
})

describe('DistributedCircuitBreakerClient', () => {
  const g: any = global
  beforeEach(() => {
    g.fetch = jest.fn().mockResolvedValue({
      json: jest.fn().mockResolvedValue({ ok: true }),
    })
  })
  afterEach(() => {
    jest.clearAllMocks()
    jest.useRealTimers()
  })

  it('register sends registration payload', async () => {
    const client = new DistributedCircuitBreakerClient('http://coord')
    const breaker = new CircuitBreaker('svc-register', { failureThreshold: 7, successThreshold: 4 })
    client.register(breaker)

    await Promise.resolve()

    expect(g.fetch).toHaveBeenCalled()
    const [url, init] = (g.fetch as jest.Mock).mock.calls[0]
    expect(url).toBe('http://coord/circuit-breakers/register')
    const body = JSON.parse((init as any).body)
    expect(body.service).toBe('svc-register')
    expect(body.failure_threshold).toBe(7)
    expect(body.success_threshold).toBe(4)
    expect((init as any).method).toBe('POST')
  })

  it('getAggregatedState returns parsed data on success', async () => {
    const aggregated = {
      service: 'svc-agg',
      consensusState: CircuitState.OPEN,
      totalNodes: 3,
      healthScore: 0.8,
      nodeStates: { a: CircuitState.OPEN, b: CircuitState.CLOSED },
    }
    ;(global as any).fetch = jest.fn().mockResolvedValue({
      json: jest.fn().mockResolvedValue(aggregated),
    })

    const client = new DistributedCircuitBreakerClient('http://coord')
    const result = await client.getAggregatedState('svc-agg')
    expect(result).toEqual(aggregated)
  })

  it('getAggregatedState returns default on error', async () => {
    ;(global as any).fetch = jest.fn().mockRejectedValue(new Error('network'))
    const client = new DistributedCircuitBreakerClient('http://coord')
    const result = await client.getAggregatedState('svc-default')
    expect(result.service).toBe('svc-default')
    expect(result.consensusState).toBe(CircuitState.CLOSED)
    expect(result.totalNodes).toBe(0)
    expect(result.nodeStates).toEqual({})
  })

  it('startSync reports state at intervals and stopSync stops reporting', async () => {
    jest.useFakeTimers()
    const fetchMock = jest.fn().mockResolvedValue({ json: jest.fn() })
    ;(global as any).fetch = fetchMock

    const client = new DistributedCircuitBreakerClient('http://coord', 1000)
    const b1 = new CircuitBreaker('svc-sync-1')
    const b2 = new CircuitBreaker('svc-sync-2')
    client.register(b1)
    client.register(b2)
    await Promise.resolve()
    fetchMock.mockClear()

    client.startSync()

    jest.advanceTimersByTime(1000)
    expect(fetchMock).toHaveBeenCalled()
    const stateCalls1 = fetchMock.mock.calls.filter(
      (c: any[]) => typeof c[0] === 'string' && c[0].includes('/circuit-breakers/state')
    )
    expect(stateCalls1.length).toBe(2)

    fetchMock.mockClear()
    jest.advanceTimersByTime(1000)
    const stateCalls2 = fetchMock.mock.calls.filter(
      (c: any[]) => typeof c[0] === 'string' && c[0].includes('/circuit-breakers/state')
    )
    expect(stateCalls2.length).toBe(2)

    fetchMock.mockClear()
    client.stopSync()
    jest.advanceTimersByTime(2000)
    expect(fetchMock).not.toHaveBeenCalled()
  })
})

describe('withCircuitBreaker decorator', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('wraps method to execute via CircuitBreaker and returns original result', async () => {
    const executeMock = jest.fn(async (fn: () => any) => fn())
    const getOrCreateSpy = jest.spyOn(CircuitBreaker, 'getOrCreate').mockReturnValue({
      name: 'decorated',
      execute: executeMock as any,
      executeSync: jest.fn(),
      getState: jest.fn(),
      getHealthInfo: jest.fn(),
    } as unknown as CircuitBreaker)

    class Service {
      value = 10
    }
    const svc = new Service()
    const target: any = svc
    Object.defineProperty(target, 'sum', {
      value: function (a: number, b: number) {
        return a + b + this.value
      },
      writable: true,
      configurable: true,
    })

    const decorator = withCircuitBreaker('decorated')
    const desc = Object.getOwnPropertyDescriptor(target, 'sum')!
    const newDesc = decorator(target, 'sum', desc)!
    Object.defineProperty(target, 'sum', newDesc)

    const result = await (target.sum as any)(1, 2)
    expect(result).toBe(13)
    expect(getOrCreateSpy).toHaveBeenCalledWith('decorated', undefined)
    expect(executeMock).toHaveBeenCalled()
  })

  it('propagates errors when original method throws inside decorator', async () => {
    const executeMock = jest.fn(async (fn: () => any) => fn())
    jest.spyOn(CircuitBreaker, 'getOrCreate').mockReturnValue({
      name: 'decorated-error',
      execute: executeMock as any,
      executeSync: jest.fn(),
      getState: jest.fn(),
      getHealthInfo: jest.fn(),
    } as unknown as CircuitBreaker)

    const target: any = {}
    Object.defineProperty(target, 'fail', {
      value: function () {
        throw new Error('method-failed')
      },
      writable: true,
      configurable: true,
    })

    const decorator = withCircuitBreaker('decorated-error')
    const desc = Object.getOwnPropertyDescriptor(target, 'fail')!
    const newDesc = decorator(target, 'fail', desc)!
    Object.defineProperty(target, 'fail', newDesc)

    await expect((target.fail as any)()).rejects.toThrow('method-failed')
  })
})

File: python-service/tests/circuit_breaker_test.py
import json
import time
import threading
import pytest
from unittest.mock import Mock, patch, call

from src.circuit_breaker import (
    CircuitState,
    CircuitBreakerConfig,
    CircuitBreakerMetrics,
    CircuitBreakerOpenError,
    CircuitBreaker,
    DistributedCircuitBreakerCoordinator,
    circuit_breaker,
)


@pytest.fixture(autouse=True)
def reset_registry():
    """Reset the CircuitBreaker registry before each test to avoid cross-test state."""
    original = CircuitBreaker._registry.copy()
    CircuitBreaker._registry.clear()
    try:
        yield
    finally:
        CircuitBreaker._registry.clear()
        CircuitBreaker._registry.update(original)


@pytest.fixture
def small_config():
    """Provide a small-threshold config for faster tests."""
    return CircuitBreakerConfig(
        failure_threshold=2,
        success_threshold=2,
        timeout_seconds=1.0,
        half_open_max_calls=2,
        sliding_window_size=4,
        failure_rate_threshold=0.5,
    )


@pytest.fixture
def breaker(small_config):
    """Create a CircuitBreaker instance with small thresholds."""
    return CircuitBreaker("test-service", config=small_config)


@pytest.fixture
def coordinator():
    """Create a DistributedCircuitBreakerCoordinator with a dummy URL."""
    return DistributedCircuitBreakerCoordinator("http://coordinator", sync_interval=0.01)


def test_circuit_state_values():
    """Test that CircuitState enum has expected values."""
    assert CircuitState.CLOSED.value == "CLOSED"
    assert CircuitState.OPEN.value == "OPEN"
    assert CircuitState.HALF_OPEN.value == "HALF_OPEN"


def test_circuit_breaker_config_defaults():
    """Test CircuitBreakerConfig default values."""
    cfg = CircuitBreakerConfig()
    assert cfg.failure_threshold == 5
    assert cfg.success_threshold == 3
    assert cfg.timeout_seconds == pytest.approx(30.0)
    assert cfg.half_open_max_calls == 3
    assert cfg.sliding_window_size == 10
    assert cfg.failure_rate_threshold == pytest.approx(0.5)


def test_metrics_record_response_time_average():
    """Test CircuitBreakerMetrics.record_response_time computes running average."""
    metrics = CircuitBreakerMetrics()
    metrics.record_response_time(0.1)
    metrics.record_response_time(0.3)
    assert metrics.average_response_time == pytest.approx(0.2)


def test_circuit_breaker_get_or_create_singleton_and_config():
    """Test get_or_create returns same instance and preserves initial config."""
    cfg = CircuitBreakerConfig(failure_threshold=9)
    br1 = CircuitBreaker.get_or_create("svc", config=cfg)
    br2 = CircuitBreaker.get_or_create("svc")
    assert br1 is br2
    assert br1.config.failure_threshold == 9


def test_circuit_breaker_state_open_to_half_open_after_timeout(breaker, monkeypatch):
    """Test state property transitions from OPEN to HALF_OPEN after timeout elapses."""
    base = 1000.0
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base)
    breaker._transition_to(CircuitState.OPEN)
    assert breaker._opened_at == pytest.approx(base)

    # After timeout_seconds elapsed, accessing state should transition to HALF_OPEN
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base + breaker.config.timeout_seconds + 0.1)
    transitions_before = breaker.metrics.state_transitions
    assert breaker.state == CircuitState.HALF_OPEN
    assert breaker.metrics.state_transitions == transitions_before + 1


def test_circuit_breaker_should_attempt_reset(breaker, monkeypatch):
    """Test _should_attempt_reset returns True only after timeout has passed."""
    assert breaker._should_attempt_reset() is False
    base = 2000.0
    breaker._opened_at = base
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base + breaker.config.timeout_seconds - 0.5)
    assert breaker._should_attempt_reset() is False
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base + breaker.config.timeout_seconds + 0.01)
    assert breaker._should_attempt_reset() is True


def test_circuit_breaker_transition_to_resets_and_sets_fields(breaker, monkeypatch):
    """Test _transition_to behavior for OPEN, HALF_OPEN, and CLOSED states."""
    base = 3000.0
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base)
    # Prepare some state
    breaker._failure_count = 5
    breaker._success_count = 4
    breaker._half_open_calls = 2
    breaker._sliding_window.extend([True, False, True])

    breaker._transition_to(CircuitState.CLOSED)
    assert breaker._failure_count == 0
    assert breaker._success_count == 0
    assert breaker._opened_at is None
    assert len(breaker._sliding_window) == 0

    breaker._transition_to(CircuitState.OPEN)
    assert breaker._opened_at == pytest.approx(base)

    breaker._transition_to(CircuitState.HALF_OPEN)
    assert breaker._half_open_calls == 0
    assert breaker._success_count == 0


def test_circuit_breaker_execute_success_and_metrics(breaker):
    """Test execute success path increments metrics and records response time."""
    result = breaker.execute(lambda: "ok")
    assert result == "ok"
    assert breaker.metrics.total_calls == 1
    assert breaker.metrics.successful_calls == 1
    assert breaker.metrics.failed_calls == 0
    assert breaker.metrics.average_response_time >= 0.0


def test_circuit_breaker_execute_failure_increments_and_raises(breaker):
    """Test execute failure path increments failure metrics and re-raises exception."""
    def op():
        raise ValueError("boom")
    with pytest.raises(ValueError):
        breaker.execute(op)
    assert breaker.metrics.total_calls == 1
    assert breaker.metrics.failed_calls == 1
    assert breaker.metrics.successful_calls == 0


def test_circuit_breaker_execute_rejects_when_open_with_fallback(breaker, monkeypatch):
    """Test execute when OPEN uses fallback and increments rejected_calls without calling operation."""
    base = 4000.0
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base)
    breaker._transition_to(CircuitState.OPEN)

    op = Mock(side_effect=AssertionError("Should not be called"))
    fb = Mock(return_value="fallback")

    res = breaker.execute(op, fallback=fb)
    assert res == "fallback"
    op.assert_not_called()
    fb.assert_called_once()
    assert breaker.metrics.rejected_calls == 1
    assert breaker.metrics.total_calls == 0  # Not counted when rejected


def test_circuit_breaker_execute_rejects_when_open_without_fallback_raises_open_error(breaker, monkeypatch):
    """Test execute when OPEN raises CircuitBreakerOpenError with correct remaining time."""
    base = 5000.0
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base)
    breaker._transition_to(CircuitState.OPEN)

    # Advance time slightly but still within timeout
    now = base + 0.2
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: now)

    with pytest.raises(CircuitBreakerOpenError) as exc:
        breaker.execute(lambda: "nope")
    err = exc.value
    assert err.name == breaker.name
    expected_remaining = breaker.config.timeout_seconds - (now - base)
    assert err.remaining_time == pytest.approx(expected_remaining, rel=1e-3, abs=1e-3)
    assert breaker.metrics.rejected_calls == 1
    assert breaker.metrics.total_calls == 0


def test_circuit_breaker_allow_request_half_open_limit_calls(breaker):
    """Test _allow_request in HALF_OPEN allows up to max calls and then rejects."""
    breaker._transition_to(CircuitState.HALF_OPEN)
    assert breaker._allow_request() is True
    assert breaker._allow_request() is True
    assert breaker._allow_request() is False
    assert breaker._half_open_calls == breaker.config.half_open_max_calls


def test_circuit_breaker_record_success_in_closed_decrements_failure_count_and_updates_metrics(breaker):
    """Test _record_success in CLOSED decrements failure_count and updates sliding window and metrics."""
    breaker._transition_to(CircuitState.CLOSED)
    breaker._failure_count = 2
    breaker._record_success(0.1)
    assert breaker._failure_count == 1
    breaker._record_success(0.3)
    assert breaker._failure_count == 0
    breaker._record_success(0.5)
    assert breaker._failure_count == 0  # Not below zero
    assert list(breaker._sliding_window)[-1] is True
    assert breaker.metrics.successful_calls == 3
    assert breaker.metrics.average_response_time == pytest.approx((0.1 + 0.3 + 0.5) / 3.0)


def test_circuit_breaker_record_success_in_half_open_closes_after_threshold(breaker):
    """Test _record_success in HALF_OPEN closes breaker after reaching success_threshold."""
    breaker._transition_to(CircuitState.HALF_OPEN)
    breaker._record_success(0.05)
    assert breaker.state == CircuitState.HALF_OPEN
    breaker._record_success(0.07)
    assert breaker.state == CircuitState.CLOSED


def test_circuit_breaker_record_failure_in_half_open_opens_immediately(breaker):
    """Test _record_failure transitions to OPEN immediately when in HALF_OPEN."""
    breaker._transition_to(CircuitState.HALF_OPEN)
    breaker._record_failure(0.02)
    assert breaker.state == CircuitState.OPEN


def test_circuit_breaker_record_failure_in_closed_opens_on_threshold(breaker):
    """Test _record_failure in CLOSED opens breaker after reaching failure_threshold."""
    breaker._transition_to(CircuitState.CLOSED)
    breaker._record_failure(0.01)
    assert breaker.state == CircuitState.CLOSED
    breaker._record_failure(0.01)
    assert breaker.state == CircuitState.OPEN
    assert breaker._opened_at is not None


def test_circuit_breaker_record_failure_opens_on_failure_rate_threshold_only():
    """Test breaker opens due to failure rate threshold when sliding window is full."""
    cfg = CircuitBreakerConfig(
        failure_threshold=100,  # effectively disable count threshold
        success_threshold=2,
        timeout_seconds=5.0,
        half_open_max_calls=2,
        sliding_window_size=4,
        failure_rate_threshold=0.5,
    )
    br = CircuitBreaker("rate-test", config=cfg)
    br._transition_to(CircuitState.CLOSED)

    # Fill sliding window with failures to exceed threshold on the 4th event
    br._record_failure(0.01)
    assert br.state == CircuitState.CLOSED  # not enough window yet
    br._record_failure(0.01)
    assert br.state == CircuitState.CLOSED
    br._record_failure(0.01)
    assert br.state == CircuitState.CLOSED
    br._record_failure(0.01)
    assert br.state == CircuitState.OPEN


def test_circuit_breaker_calculate_failure_rate(breaker):
    """Test _calculate_failure_rate returns 0.0 when window not full and correct ratio when full."""
    # Less than window size -> 0.0
    breaker._sliding_window.extend([True, False])
    assert breaker._calculate_failure_rate() == pytest.approx(0.0)

    # Fill to window size
    breaker._sliding_window.clear()
    breaker._sliding_window.extend([True, False, True, False])  # 2 failures out of 4 -> 0.5
    assert breaker._calculate_failure_rate() == pytest.approx(0.5)


def test_circuit_breaker_get_health_info_contains_expected_fields(breaker):
    """Test get_health_info returns expected structure and computes averages in ms."""
    # Ensure sliding window is full for failure_rate calculation
    breaker._transition_to(CircuitState.CLOSED)
    breaker._record_success(0.2)  # True
    breaker._record_failure(0.4)  # False
    breaker._record_success(0.6)  # True
    breaker._record_failure(0.8)  # False

    info = breaker.get_health_info()
    assert info["name"] == breaker.name
    assert info["state"] in {"CLOSED", "OPEN", "HALF_OPEN"}
    assert "failure_count" in info
    assert "success_count" in info
    assert info["failure_rate"] == pytest.approx(0.5)
    assert "metrics" in info and "config" in info

    metrics = info["metrics"]
    # Average of durations: (0.2 + 0.4 + 0.6 + 0.8) / 4 = 0.5 seconds => 500 ms
    assert metrics["average_response_time_ms"] == pytest.approx(500.0, abs=1e-6)


def test_circuit_breaker_execute_half_open_flow(breaker, monkeypatch):
    """Integration test: open -> half-open after timeout -> close after success_threshold successes."""
    # Configure to open on first failure
    breaker.config.failure_threshold = 1
    base = 6000.0
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base)

    # First failing call opens the circuit
    with pytest.raises(RuntimeError):
        breaker.execute(lambda: (_ for _ in ()).throw(RuntimeError("fail")))
    assert breaker.state == CircuitState.OPEN

    # Advance time to trigger HALF_OPEN on next check
    monkeypatch.setattr("src.circuit_breaker.time.time", lambda: base + breaker.config.timeout_seconds + 0.1)
    assert breaker.state == CircuitState.HALF_OPEN

    # Two successful trial calls -> should close
    assert breaker.execute(lambda: "ok") == "ok"
    assert breaker.state == CircuitState.HALF_OPEN
    assert breaker.execute(lambda: "ok") == "ok"
    assert breaker.state == CircuitState.CLOSED


def test_coordinator_register_breaker_sends_registration(coordinator, breaker):
    """Test that register_breaker sends a registration HTTP request."""
    with patch("src.circuit_breaker.urllib.request.urlopen") as mock_urlopen:
        mock_urlopen.return_value = Mock()
        coordinator.register_breaker(breaker)
        assert mock_urlopen.call_count == 1
        req = mock_urlopen.call_args[0][0]
        assert "/circuit-breakers/register" in req.full_url
        data = json.loads(req.data.decode("utf-8"))
        assert data["service"] == breaker.name
        assert "node_id" in data


def test_coordinator_register_breaker_handles_urlerror(coordinator, breaker):
    """Test that register_breaker suppresses URLError exceptions."""
    with patch("src.circuit_breaker.urllib.request.urlopen", side_effect=Exception("should not leak")):
        # Should not raise
        coordinator.register_breaker(breaker)


def test_coordinator_synchronize_states_posts_state(coordinator, breaker):
    """Test _synchronize_states posts state updates for each registered breaker."""
    other = CircuitBreaker("other", config=breaker.config)
    coordinator.register_breaker(breaker)
    coordinator.register_breaker(other)

    with patch("src.circuit_breaker.urllib.request.urlopen") as mock_urlopen:
        mock_urlopen.return_value = Mock()
        coordinator._synchronize_states()
        assert mock_urlopen.call_count == 2
        urls = [mock_urlopen.call_args_list[0][0][0].full_url, mock_urlopen.call_args_list[1][0][0].full_url]
        assert all("/circuit-breakers/state" in u for u in urls)
        # Validate payload structure
        for call_args in mock_urlopen.call_args_list:
            req = call_args[0][0]
            payload = json.loads(req.data.decode("utf-8"))
            assert "service" in payload
            assert "node_id" in payload
            assert "state" in payload
            assert "failure_count" in payload
            assert "timestamp" in payload
            assert "health_info" in payload


def test_coordinator_start_and_stop_sync_runs_loop(coordinator, breaker, monkeypatch):
    """Test start_sync starts thread and stop_sync stops it; loop calls _synchronize_states."""
    coordinator.register_breaker(breaker)
    sync_mock = Mock()
    monkeypatch.setattr(coordinator, "_synchronize_states", sync_mock)
    coordinator.start_sync()
    time.sleep(0.03)
    coordinator.stop_sync()
    assert sync_mock.call_count >= 1
    assert not coordinator._running
    if coordinator._sync_thread:
        assert not coordinator._sync_thread.is_alive()


def test_coordinator_get_cluster_state_success(coordinator):
    """Test get_cluster_state returns parsed JSON on success."""
    response_data = {"state": "aggregate"}
    mock_resp = Mock()
    mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")

    with patch("src.circuit_breaker.urllib.request.urlopen", return_value=mock_resp):
        result = coordinator.get_cluster_state("svc")
        assert result == response_data


def test_coordinator_get_cluster_state_urlerror(coordinator):
    """Test get_cluster_state returns error dict on URLError."""
    with patch("src.circuit_breaker.urllib.request.urlopen", side_effect=Exception("network")):
        res = coordinator.get_cluster_state("svc")
        assert res["error"] == "Failed to fetch cluster state"


def test_decorator_circuit_breaker_executes_and_sets_attributes():
    """Test circuit_breaker decorator executes function and exposes breaker."""
    @circuit_breaker("decorated-service")
    def f(x, y):
        return x + y

    res = f(2, 3)
    assert res == 5
    assert hasattr(f, "circuit_breaker")
    br = f.circuit_breaker
    assert isinstance(br, CircuitBreaker)
    assert br.metrics.total_calls == 1
    assert br.metrics.successful_calls == 1